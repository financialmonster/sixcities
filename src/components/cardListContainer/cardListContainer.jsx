import React, {useEffect} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';

import Spinner from '../spinner/spinner.jsx';
import {hasOffers, hotelsByCity} from '../../selectors.js';
import fetchOffers from '../../actions/offers.js';
import CardList from '../cardList/cardList.jsx';
import CardListEmpty from '../cardListEmpty/cardListEmpty.jsx';
import ErrorIndicator from '../errorIndicator/errorIndicator.jsx';

const CardListContainer = ({fetchOffers, offers, isLoading, activeCity, isError}) => {
    useEffect(() => {
        fetchOffers();
    }, [fetchOffers]);

    if(isLoading) {
        return <Spinner />;
    }

    if(isError) {
        return <ErrorIndicator operation="loading of the list of offers" />;
    }

    if(!offers.length) {
        return <CardListEmpty city={activeCity} />;
    }

    return <CardList offers={offers} />;
}

CardListContainer.propTypes = {
    offers: PropTypes.arrayOf(PropTypes.shape({
        price: PropTypes.number.isRequired,
        is_favorite: PropTypes.bool.isRequired,
        is_premium: PropTypes.bool.isRequired,
        rating: PropTypes.number.isRequired,
        title: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired,
        preview_image: PropTypes.string.isRequired,
        id: PropTypes.number.isRequired
    })),
    isLoading: PropTypes.bool.isRequired,
    hasOffers: PropTypes.bool.isRequired,
    activeCity: PropTypes.string,
    fetchOffers: PropTypes.func.isRequired
}

const mapStateToProps = ({offers}) => {
    return {
        offers: hotelsByCity(offers),
        isLoading: offers.isLoading,
        hasOffers: hasOffers(offers),
        activeCity: offers.activeCity,
        isError: offers.isError
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        fetchOffers: bindActionCreators(fetchOffers, dispatch)
    }
}

export {CardListContainer};
export default connect(mapStateToProps, mapDispatchToProps)(CardListContainer);
