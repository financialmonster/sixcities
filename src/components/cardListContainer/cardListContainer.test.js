import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import {BrowserRouter as Router} from 'react-router-dom';

import {CardListContainer} from './cardListContainer.jsx';

const offers =  [{price: 120, is_premium: true, is_favorite: false, rating: 4, src: `img/apartment-01.jpg`,
    title: `Beautiful luxurious apartment at great location`, type: `Apartment`,
    preview_image: `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/9.jpg`, id: 0},
    {price: 100, is_premium: false, is_favorite: true, rating: 5, src: `img/apartment-02.jpg`,
    title: `Perfectly located Castro`, type: `Room`,
    preview_image: `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/4.jpg`, id: 1},
    {price: 80, is_premium: true, is_favorite: true, rating: 3, src: `img/apartment-03.jpg`,
    title: `Waterfront with extraordinary view`, type: `Hotel`,
    preview_image: `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/3.jpg`, id: 2}
];

describe(`Correctly renders the CarsListContainer component`, () => {
    it(`when isLoading property is equal true`, () => {
        const mockStore = configureStore([]);
        const store = mockStore({offers});

        const tree = renderer.create(
            <Provider store={store}>
                <Router>
                    <CardListContainer offers={offers} fetchOffers={() => {}} activeCity={null} isError={false}
                        isLoading={true}
                    />
                </Router>
            </Provider>
            )
        .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it(`when isError property is equal true`, () => {
        const mockStore = configureStore([]);
        const store = mockStore({offers});

        const tree = renderer.create(
            <Provider store={store}>
                <Router>
                    <CardListContainer offers={offers} fetchOffers={() => {}} activeCity='Paris' isError={true}
                        isLoading={false}
                    />
                </Router>
            </Provider>
            )
        .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it(`when isError and isLoading properties are equal false and offers.length property is equal 0`, () => {
        const mockStore = configureStore([]);
        const store = mockStore({offers});

        const tree = renderer.create(
            <Provider store={store}>
                <Router>
                    <CardListContainer offers={[]} fetchOffers={() => {}} activeCity='Paris' isError={false}
                        isLoading={false}
                    />
                </Router>
            </Provider>
            )
        .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it(`when isError and isLoading properties are equal false and offers.length property is not equal 0`, () => {
        const mockStore = configureStore([]);
        const store = mockStore({offers});

        const tree = renderer.create(
            <Provider store={store}>
                <Router>
                    <CardListContainer offers={offers} fetchOffers={() => {}} activeCity='Paris' isError={false}
                        isLoading={false}
                    />
                </Router>
            </Provider>
            )
        .toJSON();

        expect(tree).toMatchSnapshot();
    });
});
