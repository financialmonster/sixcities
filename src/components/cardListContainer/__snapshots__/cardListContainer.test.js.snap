// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly renders the CarsListContainer component when isError and isLoading properties are equal false and offers.length property is equal 0 1`] = `
<div
  className="cities__status-wrapper tabs__content"
>
  <b
    className="cities__status"
  >
    No places to stay available
  </b>
  <p
    className="cities__status-description"
  >
    We could not find any property available at the moment in 
    Paris
  </p>
</div>
`;

exports[`Correctly renders the CarsListContainer component when isError and isLoading properties are equal false and offers.length property is not equal 0 1`] = `
<div
  className="cities__places-list places__list tabs__content"
>
  <a
    href="/offer/1"
    onClick={[Function]}
    title="To the Beautiful luxurious apartment at great location offer page"
  >
    <article
      className="place-card cities__place-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="place-card__mark"
      >
        <span>
          Premium
        </span>
      </div>
      <div
        className="place-card__image-wrapper cities__image-wrapper"
      >
        <img
          alt="Place pic"
          className="place-card__image"
          height="200"
          src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/9.jpg"
          width="260"
        />
      </div>
      <div
        className="place-card__info"
      >
        <div
          className="place-card__price-wrapper"
        >
          <div
            className="place-card__price"
          >
            <b
              className="place-card__price-value"
            >
              €
              120
            </b>
            <span
              className="place-card__price-text"
            >
              / night
            </span>
          </div>
          <button
            className="place-card__bookmark-button button"
            title="To bookmarks"
            type="button"
          >
            <svg
              className="place-card__bookmark-icon"
              height="19"
              width="18"
            >
              <use
                xlinkHref="#icon-bookmark"
              />
            </svg>
            <span
              className="visually-hidden"
            >
              Not in bookmarks
            </span>
          </button>
        </div>
        <div
          className="place-card__rating rating"
        >
          <div
            className="place-card__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "80%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <h2
          className="place-card__name"
        >
          Beautiful luxurious apartment at great location
        </h2>
        <p
          className="place-card__type"
        >
          Apartment
        </p>
      </div>
    </article>
  </a>
  <a
    href="/offer/2"
    onClick={[Function]}
    title="To the Perfectly located Castro offer page"
  >
    <article
      className="place-card cities__place-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="place-card__image-wrapper cities__image-wrapper"
      >
        <img
          alt="Place pic"
          className="place-card__image"
          height="200"
          src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/4.jpg"
          width="260"
        />
      </div>
      <div
        className="place-card__info"
      >
        <div
          className="place-card__price-wrapper"
        >
          <div
            className="place-card__price"
          >
            <b
              className="place-card__price-value"
            >
              €
              100
            </b>
            <span
              className="place-card__price-text"
            >
              / night
            </span>
          </div>
          <button
            className="place-card__bookmark-button button place-card__bookmark-button--active"
            title="In bookmarks"
            type="button"
          >
            <svg
              className="place-card__bookmark-icon"
              height="19"
              width="18"
            >
              <use
                xlinkHref="#icon-bookmark"
              />
            </svg>
            <span
              className="visually-hidden"
            >
              In bookmarks
            </span>
          </button>
        </div>
        <div
          className="place-card__rating rating"
        >
          <div
            className="place-card__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "100%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <h2
          className="place-card__name"
        >
          Perfectly located Castro
        </h2>
        <p
          className="place-card__type"
        >
          Room
        </p>
      </div>
    </article>
  </a>
  <a
    href="/offer/3"
    onClick={[Function]}
    title="To the Waterfront with extraordinary view offer page"
  >
    <article
      className="place-card cities__place-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="place-card__mark"
      >
        <span>
          Premium
        </span>
      </div>
      <div
        className="place-card__image-wrapper cities__image-wrapper"
      >
        <img
          alt="Place pic"
          className="place-card__image"
          height="200"
          src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/3.jpg"
          width="260"
        />
      </div>
      <div
        className="place-card__info"
      >
        <div
          className="place-card__price-wrapper"
        >
          <div
            className="place-card__price"
          >
            <b
              className="place-card__price-value"
            >
              €
              80
            </b>
            <span
              className="place-card__price-text"
            >
              / night
            </span>
          </div>
          <button
            className="place-card__bookmark-button button place-card__bookmark-button--active"
            title="In bookmarks"
            type="button"
          >
            <svg
              className="place-card__bookmark-icon"
              height="19"
              width="18"
            >
              <use
                xlinkHref="#icon-bookmark"
              />
            </svg>
            <span
              className="visually-hidden"
            >
              In bookmarks
            </span>
          </button>
        </div>
        <div
          className="place-card__rating rating"
        >
          <div
            className="place-card__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "60%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <h2
          className="place-card__name"
        >
          Waterfront with extraordinary view
        </h2>
        <p
          className="place-card__type"
        >
          Hotel
        </p>
      </div>
    </article>
  </a>
</div>
`;

exports[`Correctly renders the CarsListContainer component when isError property is equal true 1`] = `
<div
  className="cities__status-wrapper tabs__content"
>
  <b
    className="cities__status"
  >
    We're sorry
  </b>
  <p
    className="cities__status-description"
  >
    It's been an error during the process of 
    loading of the list of offers
    . Please, reload the page and try again
  </p>
</div>
`;

exports[`Correctly renders the CarsListContainer component when isLoading property is equal true 1`] = `
<div
  className="spinner"
>
  <div
    className="spinnerContent"
  >
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
  </div>
</div>
`;
